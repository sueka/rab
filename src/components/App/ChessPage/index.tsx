import React from 'react'

import Chessboard from '~/components/Chessboard'

const ChessPage: React.FunctionComponent = () => (
  <>
    <Chessboard />
  </>
)

export default ChessPage
