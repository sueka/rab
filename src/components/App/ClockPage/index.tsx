import React from 'react'
import Helmet from 'react-helmet'
import { useIntl } from 'react-intl'

import AnalogClock from '~/components/AnalogClock'
import { createPage } from '~/templates/PageTemplate'
import messages from './messages'

const ClockPage: React.FC = () => {
  const { formatMessage } = useIntl()

  return (
    <>
      <Helmet title={ formatMessage(messages.clock) } />
      <AnalogClock />
    </>
  )
}

export default createPage(ClockPage)
