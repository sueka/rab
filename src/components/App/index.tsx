import { hot } from 'react-hot-loader'
import React from 'react'
import { Switch } from 'react-router'
import Helmet from 'react-helmet'
import { InjectedIntlProps, injectIntl } from 'react-intl'

import CssBaseline from '@material-ui/core/CssBaseline'

import Route from '~/components/Route'

import messages from './messages'

export /* for testing */ const HomePage = React.lazy(() => import(/* webpackChunkName: "homePage" */ '~/components/HomePage'))
export /* for testing */ const CounterPage = React.lazy(() => import(/* webpackChunkName: "counter" */ '~/components/CounterPage'))
export /* for testing */ const InfoPage = React.lazy(() => import(/* webpackChunkName: "info" */ '~/components/InfoPage'))
export /* for testing */ const ReminderPage = React.lazy(() => import(/* webpackChunkName: "reminder" */ '~/components/ReminderPage'))
export /* for testing */ const NoMatch = React.lazy(() => import(/* webpackChunkName: "noMatch" */ '~/components/NoMatch'))

type Props =
  & InjectedIntlProps

const App: React.FunctionComponent<Props> = ({ intl: { formatMessage } }) => (
  <>
    <CssBaseline />
    <Helmet
      titleTemplate="%s - react-app-prototype"
      defaultTitle="react-app-prototype"
    />
    <Switch>
      <Route path="/" component={ HomePage } helmetProps={ { title: formatMessage(messages.home) } } />
      <Route path="/counter" component={ CounterPage } helmetProps={ { title: formatMessage(messages.counter) } } />
      <Route path="/info" component={ InfoPage } helmetProps={ { title: formatMessage(messages.info) } } />
      <Route path="/reminder" component={ ReminderPage } helmetProps={ { title: formatMessage(messages.reminder) } } />
      <Route path="*" component={ NoMatch } />
    </Switch>
  </>
)

export default hot(module)(injectIntl(App))
