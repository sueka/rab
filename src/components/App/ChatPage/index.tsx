import { IconButton, InputAdornment, TextField } from '@material-ui/core'
import { Send as SendIcon } from '@material-ui/icons'
import { useInjection } from 'inversify-react'
import React, { useCallback, useMemo, useState } from 'react'
import Helmet from 'react-helmet'
import { useIntl } from 'react-intl'

import { createPage } from '~/components/PageTemplate'
import ConfigRegistry from '~/config/ConfigRegistry'
import messages from './messages'

const ChatPage: React.FC = () => {
  const { formatMessage } = useIntl()
  const [message, setMessage] = useState('')
  const config = useInjection<ConfigRegistry>('EnvVarConfig')
  const ws = useMemo(() => new WebSocket(config.get('CHAT_SERVER_URL')), [config])

  const handleChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>((event) => {
    setMessage(event.target.value)
  }, [])

  const handleSendButtonClick = useCallback<React.MouseEventHandler>((_event) => {
    ws.send(message)
  }, [ws, message])

  return (
    <>
      <Helmet title={ formatMessage(messages.chat) } />
      <TextField
        value={ message }
        onChange={ handleChange }
        InputProps={ {
          endAdornment: (
            <InputAdornment position="end">
              <IconButton onClick={ handleSendButtonClick }>
                <SendIcon />
              </IconButton>
            </InputAdornment>
          ),
        } }
      />
    </>
  )
}

export default createPage(ChatPage)
