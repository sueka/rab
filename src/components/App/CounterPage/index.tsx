import React from 'react'
import Helmet from 'react-helmet'
import { useIntl } from 'react-intl'

import Counter from '~/components/Counter'
import { createPage } from '~/components/PageTemplate'
import messages from './messages'

const CounterPage: React.FC = () => {
  const { formatMessage } = useIntl()

  return (
    <>
      <Helmet title={ formatMessage(messages.counter) } />
      <Counter />
    </>
  )
}

export default createPage(CounterPage)
