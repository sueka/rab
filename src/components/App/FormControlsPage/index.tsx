import InputAdornment from '@mui/material/InputAdornment'
import List from '@mui/material/List'
import ListItem from '@mui/material/ListItem'
import ListItemText from '@mui/material/ListItemText'
import TextField from '@mui/material/TextField'
import { Theme, makeStyles } from '@mui/material/styles'
import React, { useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react'
import Helmet from 'react-helmet'
import { FormattedMessage, useIntl } from 'react-intl'

import CodeField from '~/components/CodeField'
import CopyTextButton from '~/components/CopyTextButton'
import FileUpload from '~/components/FileUpload'
import fileUploadMessages from '~/components/FileUpload/messages' // TODO
import MicSwitch, { Props as MicSwitchProps } from '~/components/MicSwitch'
import { createPage } from '~/components/PageTemplate'
import IntlProviderContext from '~/contexts/IntlProviderContext'
import useScreen from '~/hooks/useScreen'
import messages from './messages'
import { shouldBePresent } from '~/asserters/commonAsserters'

const CopiableTextField: React.FC = () => {
  const [text, setText] = useState('')
  const input = useRef<HTMLInputElement>(null)

  const handleChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>((event) => {
    setText(event.target.value)
  }, [])

  return (
    <TextField
      variant="standard"
      label="label"
      value={ text }
      onChange={ handleChange }
      inputRef={ input }
      InputProps={ {
        endAdornment: (
          <InputAdornment position="end">
            <CopyTextButton inputFor={ input } />
          </InputAdornment>
        ),
      } }
    />
  )
}

const MicIncludedTextField: React.FC = () => {
  const { dir } = useContext(IntlProviderContext)
  const [text, setText] = useState('')
  const input = useRef<HTMLInputElement>(null)
  const scrollWidth = input.current?.scrollWidth
  const scrollHeight = input.current?.scrollHeight
  const offsetWidth = input.current?.offsetWidth
  const offsetHeight = input.current?.offsetHeight

  const handleChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>((event) => {
    setText(event.target.value)
  }, [])

  const handleResult = useCallback<MicSwitchProps['onResult']>((_event, value) => {
    if (input.current === null) {
      return
    }

    setText(value)

    shouldBePresent(scrollWidth)
    shouldBePresent(scrollHeight)
    shouldBePresent(offsetWidth)
    shouldBePresent(offsetHeight)

    input.current.scrollTop = scrollHeight - offsetHeight

    switch (dir) {
      case 'ltr':
        input.current.scrollLeft = scrollWidth - offsetWidth
        break
      case 'rtl':
        input.current.scrollLeft = -(scrollWidth - offsetWidth)
    }
  }, [
    dir,
    input,
    // To handle resizing, DOM updates, etc:
    scrollWidth,
    scrollHeight,
    offsetWidth,
    offsetHeight,
  ])

  return (
    <TextField
      variant="standard"
      label="label"
      value={ text }
      onChange={ handleChange }
      inputRef={ input }
      InputProps={ {
        endAdornment: (
          <InputAdornment position="end">
            <MicSwitch
              inputFor={ input }
              onResult={ handleResult }
              fallback={ null }
            />
          </InputAdornment>
        ),
      } }
    />
  )
}

const MicIncludedCopiableTextField: React.FC = () => {
  const { dir } = useContext(IntlProviderContext)
  const [text, setText] = useState('')
  const input = useRef<HTMLInputElement>(null)
  const scrollWidth = input.current?.scrollWidth
  const scrollHeight = input.current?.scrollHeight
  const offsetWidth = input.current?.offsetWidth
  const offsetHeight = input.current?.offsetHeight

  const handleChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>((event) => {
    setText(event.target.value)
  }, [])

  const handleResult = useCallback<MicSwitchProps['onResult']>((_event, value) => {
    if (input.current === null) {
      return
    }

    setText(value)

    shouldBePresent(scrollWidth)
    shouldBePresent(scrollHeight)
    shouldBePresent(offsetWidth)
    shouldBePresent(offsetHeight)

    input.current.scrollTop = scrollHeight - offsetHeight

    switch (dir) {
      case 'ltr':
        input.current.scrollLeft = scrollWidth - offsetWidth
        break
      case 'rtl':
        input.current.scrollLeft = -(scrollWidth - offsetWidth)
    }
  }, [
    dir,
    input,
    // To handle resizing, DOM updates, etc:
    scrollWidth,
    scrollHeight,
    offsetWidth,
    offsetHeight,
  ])

  return (
    <TextField
      variant="standard"
      label="label"
      value={ text }
      onChange={ handleChange }
      inputRef={ input }
      InputProps={ {
        endAdornment: (
          <InputAdornment position="end">
            <MicSwitch
              inputFor={ input }
              onResult={ handleResult }
              fallback={ null }
            />
            <CopyTextButton inputFor={ input } />
          </InputAdornment>
        ),
      } }
    />
  )
}

interface ImageFileProps {
  file: File
}

interface StyleProps {
  width: number | null
  height: number | null
}

const useStyles = makeStyles<Theme, StyleProps, 'Image'>({
  Image: {
    maxWidth: ({ width }) => width !== null ? width / 2 : undefined,
    maxHeight: ({ height }) => height !== null ? height / 2 : undefined,
  },
})

const ImageFile: React.FC<ImageFileProps> = ({ file }) => {
  const [src, setSrc] = useState<string | null>(null)
  const reader = useMemo(() => new FileReader, [])
  const { width, height } = useScreen()
  const jssClasses = useStyles({ width, height })

  const handleReaderLoad = useCallback<NonNullable<FileReader['onload']>>((event) => {
    if (event.target?.result == null) {
      return
    }

    if (event.target.result instanceof ArrayBuffer) {
      throw new Error //
    }

    setSrc(event.target.result)
  }, [])

  useEffect(() => {
    reader.addEventListener('load', handleReaderLoad)

    return () => {
      reader.removeEventListener('load', handleReaderLoad)
    }
  }, [reader, handleReaderLoad])

  useEffect(() => {
    if (!/^image\/[0-9A-Za-z][!#$&+-.0-9A-Z^_a-z]*$/.test(file.type)) {
      throw new Error
    }

    reader.readAsDataURL(file)
  }, [file, reader])

  return (
    <img src={ src ?? undefined } className={ jssClasses.Image } />
  )
}

const FormControlsPage: React.FC = () => {
  const { formatMessage } = useIntl()
  const [files, setFiles] = useState<File[] | null>(null)

  const handleFileUploadChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>((event) => {
    setFiles(event.target.files !== null ? Array.from(event.target.files) : event.target.files)
  }, [])

  const resultMessage = useMemo(() => {
    if (files === null || files.length === 0) {
      return <FormattedMessage { ...fileUploadMessages.noFileSelected } />
    }

    return Array.from(files).map((file, i) => <ImageFile key={ i } file={ file } />)
  }, [files])

  const [code, setCode] = useState('')

  const handleCodeFieldChange = useCallback<React.ChangeEventHandler<HTMLTextAreaElement>>((event) => {
    setCode(event.target?.value)
  }, [])

  const codeFieldInput = useRef<HTMLInputElement | HTMLTextAreaElement>(null)

  return (
    <>
      <Helmet title={ formatMessage(messages.formControls) } />
      <List>
        <ListItem>
          <ListItemText>
            <FileUpload
              accept="image/*"
              multiple
              resultMessage={ resultMessage }
              onChange={ handleFileUploadChange }
            />
          </ListItemText>
        </ListItem>
        <ListItem>
          <ListItemText>
            <CopiableTextField />
          </ListItemText>
        </ListItem>
        <ListItem>
          <ListItemText>
            <MicIncludedTextField />
          </ListItemText>
        </ListItem>
        <ListItem>
          <ListItemText>
            <MicIncludedCopiableTextField />
          </ListItemText>
        </ListItem>
        <ListItem>
          <ListItemText>
            <CodeField
              label="code"
              value={ code }
              onChange={ handleCodeFieldChange }
              InputProps={ {
                endAdornment: (
                  <InputAdornment position="end">
                    <CopyTextButton inputFor={ codeFieldInput } />
                  </InputAdornment>
                ),
              } }
              inputProps={ {
                ref: codeFieldInput,
              } }
            />
          </ListItemText>
        </ListItem>
      </List>
    </>
  )
}

export default createPage(FormControlsPage)
