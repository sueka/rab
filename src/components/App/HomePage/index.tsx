import InputAdornment from '@material-ui/core/InputAdornment'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import { useInjection } from 'inversify-react'
import React, { useCallback, useRef, useState } from 'react'
import Helmet from 'react-helmet'
import { FormattedNumber, useIntl } from 'react-intl'

import CurrentTimeOfDay from '~/components/CurrentTimeOfDay'
import NotifyMeButton from '~/components/NotifyMeButton'
import ObtainCookieConsentButton from '~/components/ObtainCookieConsentButton'
import { createPage } from '~/components/PageTemplate'
import SetClockButton from '~/components/SetClockButton'
import Today from '~/components/Today'
import ConfigRegistry from '~/config/ConfigRegistry'
import messages from './messages'

const HomePage: React.FC = () => {
  const config = useInjection<ConfigRegistry>('EnvVarConfig')
  const gtmContainerId = config.get('GTM_CONTAINER_ID')
  const { formatMessage } = useIntl()
  const input = useRef<HTMLInputElement>(null)
  const [text, setText] = useState('')

  const handleChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>((event) => {
    setText(event.target.value)
  }, [])

  return (
    <>
      <Helmet title={ formatMessage(messages.home) } />
      <Typography>
        <FormattedNumber format="usd" value={ 100 } />
      </Typography>
      <Typography>
        <Today /> <CurrentTimeOfDay />
      </Typography>
      <SetClockButton />
      <TextField
        value={ text }
        onChange={ handleChange }
        InputProps={ {
          endAdornment: (
            <InputAdornment position="end">
              <NotifyMeButton inputFor={ input } />
            </InputAdornment>
          ),
        } }
        inputProps= { {
          ref: input,
        } }
      />
      { gtmContainerId !== undefined && <ObtainCookieConsentButton /> }
    </>
  )
}

export default createPage(HomePage)
