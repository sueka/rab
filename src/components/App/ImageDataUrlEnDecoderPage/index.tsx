import Box from '@mui/material/Box'
import Grid from '@mui/material/Grid'
import Typography from '@mui/material/Typography'
import React from 'react'
import Helmet from 'react-helmet'
import { FormattedMessage, useIntl } from 'react-intl'

import { createPage } from '~/components/PageTemplate'
import ImageDataUrlDecoder from '~/components/ImageDataUrlDecoder'
import ImageDataUrlEncoder from '~/components/ImageDataUrlEncoder'
import messages from './messages'

const ImageDataUrlEnDecoderPage: React.FC = () => {
  const { formatMessage } = useIntl()

  return (
    <>
      <Helmet title={ formatMessage(messages.imageDataUrlEnDecoder) } />
      <Box>
        <Box padding={ 2 }>
          <Grid container spacing={ 2 }>
            <Grid item xs={ 12 }>
              <Typography variant="h3">
                <FormattedMessage { ...messages.encoder } />
              </Typography>
            </Grid>
            <Grid item xs={ 12 }>
              <ImageDataUrlEncoder />
            </Grid>
          </Grid>
        </Box>
        <Box padding={ 2 }>
          <Grid container spacing={ 2 }>
            <Grid item xs={ 12 }>
              <Typography variant="h3">
                <FormattedMessage { ...messages.decoder } />
              </Typography>
            </Grid>
            <Grid item xs={ 12 }>
              <ImageDataUrlDecoder />
            </Grid>
          </Grid>
        </Box>
      </Box>
    </>
  )
}

export default createPage(ImageDataUrlEnDecoderPage)
