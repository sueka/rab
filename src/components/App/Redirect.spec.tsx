// TODO: index.spec.tsx に統合する?

import { render } from '@testing-library/react'
import { Provider as ServiceProvider } from 'inversify-react'
import React from 'react'
import reactRouter, { Redirect } from 'react-router'
import { RecoilRoot } from 'recoil'

import inversifyContainer from '~/container.dev'
import App from '.'

jest.mock('react-router', () => ({
  ...(jest.requireActual('react-router') as any),
  Redirect: jest.fn(() => null),
  useLocation: jest.fn(),
}))

const mockedReactRouter = reactRouter as jest.Mocked<typeof reactRouter>

describe('App', () => {
  describe('with #fragment', () => {
    it('should redirect to the page whose path-abempty is the fragment when the path-abempty is "/" and the fragment exists', () => {
      mockedReactRouter.useLocation.mockImplementationOnce(() => ({ pathname: '/', hash: '#/foo', search: '', state: undefined }))

      const { container } = render(
        <RecoilRoot>
          <ServiceProvider container={ inversifyContainer }>
            <App />
          </ServiceProvider>
        </RecoilRoot>
      )

      expect(container).toMatchSnapshot()
      expect(Redirect).toBeCalledTimes(1)
      expect(Redirect).toBeCalledWith({ to: { pathname: '/foo' } }, {})
    })
  })
})
