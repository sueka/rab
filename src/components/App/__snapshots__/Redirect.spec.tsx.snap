// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App with #fragment should redirect to the page whose path-abempty is the fragment when the path-abempty is "/" and the fragment does exist 1`] = `<div />`;
